### PR TITLE
Writing to control input ports w/o UI (DSP -> host; via atom output port) - RFC

### PR DESCRIPTION
**Pull request by rncbc**
As briefly discussed on #lv2 irc and devel@lists.lv2plug.in ...

This is all about tackling an ages old long and overdue problem, eg. one of LV2 instrument plugins that wish to support some kind of not-initiated-by-the-GUI-ever preset/state change: standard MIDI bank+program_change channel messages comes to mind...

Hopefully, this proposal tries to solve one controllers connundrum: plugins are not allowed to override their own input control port values--formerly regarded as parameters on LADSPA and now *legacy* for LV2--which are, as ever were *read-only*, of course: ownership and value is host's prerogative and sole responsability...

So, how do we make a host to update on over a plugin's control input ports you ask? 

Until now that was not possible at all; of course, control input ports are read-only from the plugin's POV; only the host may write to it on its own premises.

This rather simple protocol being here proposed should help on letting the plugin tell the host that it wishes to update or write to one or more input control ports--the host may or may not concur on its own.

And how does the plugin tells so? It does it through one of its atom output ports(*), in the very same and similar fashion like it used to do with patch properties/parameters all the way down: via an LV2 Atom object that is.

TL;DR: This proposal boils down to add just a couple of new URIs to the LV2 reportoire (possibly to `lv2/lv2plug.in/ns/ext/atom/atom.h`):
```C
#define LV2_ATOM__PortEvent LV2_ATOM_PREFIX "PortEvent"
#define LV2_ATOM__portTuple LV2_ATOM_PREFIX "portTuple"
```
Just not sure about namenclature: should it be under the `LV2:Atom` class, like `LV2_ATOM__PortEvent` suggested above? or `LV2_ATOM__PortWrite`? Or should it belong to `LV2:Patch` realm (in `lv2/lv2plug.in/ns/ext/patch/patch.h`) as `LV2_PATCH__PortEvent`, considering that it's similar in semantics to `LV2_PATCH__Put` for LV2:Patch properties/parameters but applying to control input ports instead?

Ontology and design specification looks fine to me as is and the good news are that it's implemented already on yours truly **qtractor** (as host) and the *Vee-One-Suite*, **synthv1**, **samplv1**, **drumkv1** and **padthv1** (as plugins).

---
(*) plural, although having more than one atom output port is quite unusual.

Below you may find some code snippets, as example implementations to either side, host and plugin.

---
Host: ** read_port_event **
```C
#include "lv2/lv2plug.in/ns/ext/atom/util.h"

//  LV2_Atom *atom;
//
//  `atom` is a `LV2_ATOM__Object`, read from the plugin's atom output
//  port buffer, while on the real-time processing thread and then
//  transferred to the main host (non-DSP) thread, possibly via a
//  ring-buffer, wrapped in a `LV2_ATOM__eventTransfer` container.
//
    const LV2_Atom_Object *obj = (const LV2_Atom_Object *) atom;
    // ...
    if (obj->body.otype == _urid_map(LV2_ATOM__PortEvent)) {
        const LV2_Atom_Tuple *tup = nullptr;
        lv2_atom_object_get(obj,
            _urid_map(LV2_ATOM__portTuple), (const LV2_Atom *) &tup, 0);
        uint32_t port_index = 0;
        if (tup) {
            uint32_t port_index = 0;
            LV2_ATOM_TUPLE_FOREACH(tup, iter) {
                if (iter->type == _urid_map(LV2_ATOM__Int))
                    port_index = *(uint32_t *) (iter + 1);
                else
                if (iter->type == _urid_map(LV2_ATOM__Float)) {
                    const uint32_t buffer_size = iter->size;
                    const void *buffer = iter + 1;
                    _control_port_write(port_index, buffer_size, 0, buffer);
                }
            }
        }
    }
//  ...
```

---
Plugin: ** write_port_event(s) **
```C
#include "lv2/lv2plug.in/ns/ext/atom/forge.h"

//  LV2_Atom_Forge *forge;
//
//  `forge` should be bound to the plugin's atom output port buffer;
//  nb. remember to declare an adequate port buffer minimum size (via
//  `LV2_RESIZE_PORT__minimumSize` property) if you intend to notify in
//  bulk ie. more than one tuple or pairs (`port_index`, `port_value`)
//  in contiguous sequence.
//
//  uint32_t port_index;
//  float port_value;
//
//  `port_index` (int) is the control input port index that is desired
//  to get updated on the host-side with `port_value` (float), forming
//  one or more tuples wrapped in a LV2_ATOM__PortEvent object and sent
//  out to host.
//
    lv2_atom_forge_frame_time(forge, 0);

    LV2_Atom_Forge_Frame obj_frame;
    lv2_atom_forge_object(forge, &obj_frame, 0, _urid_map(LV2_ATOM__PortEvent));
    lv2_atom_forge_key(forge, _urid_map(LV2_ATOM__portTuple));

    LV2_Atom_Forge_Frame tup_frame;
    lv2_atom_forge_tuple(forge, &tup_frame);

    // for each (port_index, port_value) tuple ...
    lv2_atom_forge_int(forge, port_index);
    lv2_atom_forge_float(forge, port_value);
    // ...

    lv2_atom_forge_pop(forge, &tup_frame);
    lv2_atom_forge_pop(forge, &obj_frame);

//  ...
```

### CODE REVIEW

**Comment by sadko4u** [comment]
The only thing I don't like is... use of Atom ports for updating such simple primitives like LADSPA-derived ports. I believe it would be more rational to provide an extension interface instead like this:
```C
typedef void* LV2_PortWrite_Handle;

typedef struct LV2_PortWrite {
    /* Handle */
    LV2_PortWrite_Handle handle;

    /* Same to the LV2UI_Write_Function */
    LV2_URID (*write)(LV2_PortWrite_Handle handle, 
                                     uint32_t         port_index,
                                     uint32_t         buffer_size,
                                     uint32_t         port_protocol,
                                     const void*      buffer);
} LV2_PortWrite;
```

This extension interface restricts participants to DSP and Host only without affecting the UI which can catch the Atom event.

**Comment by falkTX** [comment]
Would be nice to get some sort of resolution here. 
A better name for the properties would be "ControlPortEvent" and "controlPortEvent", as this is only ever going to be useful for control ports, correct?

**Comment by alex-tee** [comment]
I too prefer the extension approach with a setter. in most (all?) cases it will just floats so although using atoms might seem more elegant, IMO it's over-engineering and makes it more complicated to implement

that said I agree with @drobilla on this

> in general it's a (contentious) extension that adds API, so as far as including goes, the usual rules these days apply: several separate independent implementations need to agree on it and implement it and work with each other. Until then, implementations should use third-party URIs.

I would definitely add this feature to Zrythm so you can count me in

**Comment by falkTX** [comment]
Related to this, I have created https://github.com/KXStudio/LV2-Extensions/tree/master/kx-control-input-port-change-request.lv2
Implemented in mod-host, will add it in Carla at some point soon.

**Comment by drobilla** [comment]
@falkTX I think it would be better if that had more parity with LV2UI_Request_Value

Perhaps literally being the same feature, though that would mean you need to add a designation for the control.

**Comment by falkTX** [comment]
> @falkTX I think it would be better if that had more parity with LV2UI_Request_Value
> 
> Perhaps literally being the same feature, though that would mean you need to add a designation for the control.

I thought about this, but can't come up with any other use for it. We could make it generic with a pointer and size, but since this is exactly for the case of changing input-control ports, it will not ever change right?
